### PR TITLE
fix: abac not using policy

### PR DESCRIPTION
**Pull request by Abingcbc**
Signed-off-by: abingcbc <abingcbc626@gmail.com>

Fix: https://github.com/casbin/casbin/issues/917

## How it works

Ignore all policies when matcher doesn't contain any match related to policy.

### CODE REVIEW

**Comment by hsluoyz** [comment]
@nodece plz review

**Comment by hsluoyz** [comment]
@silverspace plz review

**Review by nodece — APPROVED** [review]
LGTM

File: examples/abac_rule_deny_model.conf [hunk 1]
```diff
@@ -0,0 +1,11 @@
+[request_definition]
+r = sub, obj, act
+
+[policy_definition]
+p = sub_rule, obj, act, eft
+
+[policy_effect]
+e = some(where (p.eft == allow))
```
Comment (silverspace) [inline examples/abac_rule_deny_model.conf:8]: This is missing the critical deny policy effect. Could you change this to:
```
e = some(where (p.eft == allow)) && !some(where (p.eft == deny))
```

Comment (silverspace) [reply examples/abac_rule_deny_model.conf:8]: (I think this issue is specific to the `allow-and-deny` policy effect.)

File: enforcer.go [hunk 2]
```diff
@@ -550,7 +551,7 @@ func (e *Enforcer) enforce(matcher string, explains *[]string, rvals ...interfac
 	var effect effector.Effect
 	var explainIndex int
 
-	if policyLen := len(e.model["p"][pType].Policy); policyLen != 0 {
+	if policyLen := len(e.model["p"][pType].Policy); policyLen != 0 && strings.Contains(expString, pType + "_"){
```
Comment (silverspace) [inline enforcer.go:554]: Thanks for this super fast PR!

File: examples/abac_rule_deny_model.conf (hunk 1 above)
Comment (hsluoyz) [reply examples/abac_rule_deny_model.conf:8]: @Abingcbc

Comment (Abingcbc) [reply examples/abac_rule_deny_model.conf]: Done❤️

**Review by sagilio — CHANGES_REQUESTED** [review]
Can you provide a more reasonable test case? I try to remove the matcher check and run the test, it returns another error:
image
image

Then I try to add `eft` to the abac_model.conf, the new test is passed:
image

I think your changes have nothing to do with the error in the issue.

**Comment by Abingcbc** [comment]
> Can you provide a more reasonable test case? I try to remove the matcher check and run the test, it returns another error: image

@sagilio sa I don't understand what you mean, can you describe in more detail？

**Comment by Abingcbc** [comment]
@sagilio Yep. I realized this error when  I wrote the test. But if the matcher doesn't use policy, no policy should be matched for the request. So no matter what policy definitions look like,  policy should have no effect on the request?

**Comment by sagilio** [comment]
> @sagilio Yep. I realized this error when I wrote the test. But if the matcher doesn't use policy, no policy should be matched for the request. So no matter what policy definitions look like, policy should have no effect on the request?

Yeah, you changes is right，we need not scan.all policies when matcher is not include rtype. But this just a enhance about performance. I do not know why the wrong result returned in #917.

**Comment by Abingcbc** [comment]
> > @sagilio Yep. I realized this error when I wrote the test. But if the matcher doesn't use policy, no policy should be matched for the request. So no matter what policy definitions look like, policy should have no effect on the request?
> 
> Yeah, you changes is right，we need not scan.all policies when matcher is not include rtype. But this just a enhance about performance. I do not know why the wrong result returned in #917.

I think the root cause is that `policy` is for other model like `rbac` and the request's model is `abac`.

When the matcher doesn't use any policy and the request meets the judgment of the matcher, the request will match all policies because the matcher always returns true.
Therefore, if there is `deny` policy, the result of merged effects will be `deny`.

Take https://github.com/casbin/casbin/issues/917 as an example, request:
```
{ User: 'alice', Superuser: true }, { Owner: 'bob'}, read
```

will match policy

```
p, foobar, /blah, read, deny
```

and the merged effect is `deny`, although it is allowed in respect to `abac`.

**Comment by sagilio** [comment]
@Abingcbc I see, you are right. We may need a more effective test case to descript the situation.

**Comment by Abingcbc** [comment]
> @Abingcbc I see, you are right. We may need a more effective test case to descript the situation.

@sagilio I have updated the test case. PTAL ❤️

**Comment by hsluoyz** [comment]
@sagilio plz review

**Review by sagilio — APPROVED** [review]
LGTM

**Comment by harshalmittal4** [comment]
Hi @Abingcbc , @hsluoyz ,
I was seeking to understand the abac policy and read the discussion above and at https://github.com/casbin/casbin/issues/917, and get that the request must not depend on policy file for a model like below, i.e. it must be allowed everytime the condition `r.sub == r.obj.Owner` holds. Is this correct?

model - 

```
[request_definition]
r = sub, obj, act

[policy_definition]
p = sub, obj, act, eft

[policy_effect]
e = some(where (p.eft == allow)) && !some(where (p.eft == deny))

[matchers]
m = r.sub == r.obj.Owner
```

For a request as `alice, { Owner: 'alice'}, write`, I used the above model and the below policy to run in online editor - 
```
p, alice, /data1, read, deny
p, alice, /data1, write, allow
p, bob, /data2, write, allow
p, bob, /data2, read, allow
```
 and got the result as false. 

On replacing `p, alice, /data1, read, deny` in policy file with `p, alice, /data1, read, allow` (so that all eft in policy become allow), the result changes to true. Is this the correct behavior (i.e request depending on policy file)?

**Comment by hsluoyz** [comment]
@harshalmittal4 the online editor is pure frontend (https://github.com/casbin/casbin-editor), which uses Node-Casbin: https://github.com/casbin/node-casbin for Casbin enforcement. Usually we have latest features in Golang version Casbin and Node-Casbin maybe still doesn't have the feature of this PR: https://github.com/casbin/casbin/pull/918 yet. 

Issue proposed here: https://github.com/casbin/node-casbin/issues/433

**Comment by harshalmittal4** [comment]
Thanks @hsluoyz , got it. I'm currently getting more familiar with Casbin and would like to try contributing to some issues once I get comfortable

**Pull request merged**